src/guardian/duplicates/keys-duplication-checker.service.ts

```diff
@@ -167,37 +235,84 @@ export class KeysDuplicationCheckerService {
     return events;
   }
 
-  private findOriginalEvent(events: SigningKeyEvent[]): SigningKeyEvent {
-    return events.reduce(
-      (prev, curr) =>
-        prev.blockNumber < curr.blockNumber ||
-        (prev.blockNumber === curr.blockNumber && prev.logIndex < curr.logIndex)
-          ? prev
-          : curr,
-      events[0],
+  private getOperatorsWithoutEvents(
+    operators: string[],
+    events: SigningKeyEvent[],
+  ): string[] {
+    const eventOperators = new Set(
+      events.map((event) => `${event.moduleAddress}-${event.operatorIndex}`),
     );
+    return operators.filter((op) => !eventOperators.has(op));
   }
 
-  private findOriginalKey(
-    occurrences: RegistryKey[],
-    originalEvent: SigningKeyEvent,
-  ): RegistryKey {
-    const keyOwnerKeys = occurrences.filter(
-      (key) =>
-        key.moduleAddress === originalEvent.moduleAddress &&
-        key.operatorIndex === originalEvent.operatorIndex,
+  private filterNonEarliestKeys(
+    earliestEvent: SigningKeyEvent,
+    suspectedDuplicateKeys: RegistryKey[],
+    blockData: BlockData,
+  ) {
+    const operatorKeys = this.findOperatorKeys(
+      suspectedDuplicateKeys,
+      earliestEvent.moduleAddress,
+      earliestEvent.operatorIndex,
+    );
+
+    const earliestKey = this.findEarliestKeyWithinOperator(operatorKeys);
+
+    this.logger.log('Earliest key is', {
+      ...{
+        earliestKey,
+        createBlockNumber: earliestEvent.blockNumber,
+        createBlockHash: earliestEvent.blockHash,
+      },
+      currentBlockNumber: blockData.blockNumber,
+      currentBlockHash: blockData.blockHash,
+    });
+    return suspectedDuplicateKeys.filter(
+      (key) => !this.isSameKey(key, earliestKey),
     );
-    return this.findOriginalKeyWithinOperator(keyOwnerKeys);
   }
 
-  private findMissingOperators(
-    operators: Set<string>,
-    events: SigningKeyEvent[],
-  ): string[] {
-    const eventOperators = new Set(
-      events.map((event) => `${event.moduleAddress}-${event.operatorIndex}`),
+  private findEarliestEvents(events: SigningKeyEvent[]): SigningKeyEvent[] {
+    if (events.length <= 1) return events;
+
+    const { blockEvents } = events.reduce(
+      ({ earliestBlockNumber, blockEvents }, currEvent) => {
+        if (earliestBlockNumber === currEvent.blockNumber) {
+          blockEvents.push(currEvent);
```

Comment:
This pattern can lead to infinite loops in case of an error, it is better to avoid it. I suggested to reconsider the solution of this method in favour of Math.min and filter